src/backend/distributed/utils/causal_clock.c

```diff
@@ -0,0 +1,759 @@
+ */
+void
+InitClockAtBoot(void)
+{
+	if (creating_extension)
+	{
+		/* No catalog present yet */
+		return;
+	}
+
+	SpinLockAcquire(&logicalClockShmem->clockMutex);
+
+	/* Avoid repeated initialization */
+	if (logicalClockShmem->clockInitialized == true)
+	{
+		SpinLockRelease(&logicalClockShmem->clockMutex);
+		return;
+	}
+
+	/*
+	 * Set the flag before executing a distributed query
+	 * (else it might trigger this routine recursively.)
+	 */
+	logicalClockShmem->clockInitialized = true;
+
+	/* Start with the wall clock value */
+	uint64 epochValue = GetEpochTimeMs();
+	logicalClockShmem->clusterClockValue = START_NEW_CLOCK(epochValue);
+
+	SpinLockRelease(&logicalClockShmem->clockMutex);
+
+	/*
+	 * Version checking is a must for the below code as it accesses the
+	 * new catalog created in the new version.
+	 * Note: If we move this code to the top of the routine it will cause an
+	 * infinite loop with CitusHasBeenLoaded() calling this routine again.
+	 */
+	bool saveEnableVersionChecks = EnableVersionChecks;
+	EnableVersionChecks = true;
+
+	if (!CheckCitusVersion(NOTICE))
+	{
+		EnableVersionChecks = saveEnableVersionChecks;
+		return;
+	}
+
+	EnableVersionChecks = saveEnableVersionChecks;
+
+	/*
+	 * Select the highest clock value persisted in the catalog.
+	 *
+	 * SELECT i, MAX(j) FROM <tab>
+	 * WHERE i = (SELECT MAX(i) FROM <tab>)
+	 * GROUP BY i
+	 */
+	uint32 numCols = 2;
+	uint64 *results = ExecuteQueryAndReturnResults(
+		"SELECT clocklogical, clockcounter FROM "
+		"pg_catalog.pg_dist_commit_transaction ORDER BY 1 DESC, "
+		"2 DESC LIMIT 1;", numCols, SPI_OK_SELECT);
+
+	if (results != NULL)
+	{
+		uint64 logicalMaxClock = results[0];
+		uint32 counterMaxClock = results[1];
+
+		ereport(LOG,
+				(errmsg("Adjusted the clock with value persisted"
+						"logical(%lu) and counter(%u)",
+						logicalMaxClock, counterMaxClock)));
+
+		/*
+		 * Adjust the local clock according to the most recent
+		 * clock stamp value persisted in the catalog.
+		 */
+		AdjustLocalClockToRemote(logicalMaxClock, counterMaxClock);
+	}
+
+	/*
+	 * NULL results indicate no prior commit timestamps on this node, start
+	 * from the wall clock.
+	 */
+}
+
+
+/*
+ * ExecuteQueryAndReturnResults connects to SPI, executes the query and checks
+ * if the SPI returned the correct type. Returns an array of int64 results
+ * in the caller's memory context.
+ */
+static uint64 *
+ExecuteQueryAndReturnResults(char *query, int resultSize, int spiok_type)
+{
+	/* Allocate in caller's context */
+	uint64 *results = (uint64 *) palloc(resultSize * sizeof(uint64));
+
+	int spiResult = SPI_connect();
+	if (spiResult != SPI_OK_CONNECT)
+	{
+		ereport(ERROR, (errmsg("could not connect to SPI manager")));
+	}
+
+	spiResult = SPI_execute(query, false, 0);
+	if (spiResult != spiok_type)
+	{
+		ereport(ERROR, (errmsg("could not run SPI query")));
+	}
+
+	if (SPI_processed != 1)
+	{
+		SPI_finish();
+
+		/* Either no row or more than one(unexpected) */
```

Comment:
might be nicer if we can throw an error here

Reply:
Done.